File: 04-linked-lists/05_remove_nth.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}


ListNode* remove_nth_from_end(ListNode* head, int n) {
    return NULL;
}

int main() {
    int out[10], len;
    to_array(remove_nth_from_end(make_list(ARR(1,2,3,4,5),5), 2), out, &len);
    TEST_ARR("remove 2nd from end", out, ARR(1,2,3,5), 4);
    to_array(remove_nth_from_end(make_list(ARR(1),1), 1), out, &len);
    TEST_INT("remove only", len, 0);
    to_array(remove_nth_from_end(make_list(ARR(1,2),2), 2), out, &len);
    TEST_ARR("remove head", out, ARR(2), 1);
    TEST_SUMMARY();
}

File: 04-linked-lists/05_remove_nth.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests



# Remove Nth Node From End of List (LC #19) -- Medium
# Input:  head: ListNode, n: int (1 <= n <= list length)
# Output: ListNode (head after removing nth node from end)
# Remove the nth node from the END of the list. Return modified head.
# Guaranteed n is valid.
#
# Example:
#   1->2->3->4->5, n=2  ->  1->2->3->5  (removed 4)
#   1, n=1               ->  (empty)     (removed only node)
#   1->2, n=2            ->  2           (removed head)
#
#   Key: Two pointers with gap of n.
#        Advance fast n steps ahead. Then move both until fast hits end.
#        Slow is now just before the target. Use dummy head for edge cases.

class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


def remove_nth_from_end(head, n):
    pass


def solve(arr, n):
    return to_array(remove_nth_from_end(make_list(arr), n))


run_tests(solve, [
    (([1,2,3,4,5], 2),  [1,2,3,5]),     # remove 4th node
    (([1], 1),           []),             # remove only node
    (([1,2], 1),         [1]),            # remove tail
    (([1,2], 2),         [2]),            # remove head
    (([1,2,3], 3),       [2,3]),          # remove head of 3
    (([1,2,3,4,5], 1),  [1,2,3,4]),      # remove last
    (([1,2,3,4,5], 5),  [2,3,4,5]),      # remove first
])
